src/Compression/Deflater.cs

```diff
@@ -0,0 +1,177 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Security;
+
+using ObjCRuntime;
+
+namespace Compression
+{
+	/// <summary>
+	/// Provides a wrapper around the ZLib compression API.
+	/// </summary>
+	internal sealed class Deflater : IDisposable
+	{
+		private CompressionStreamStruct _compression_struct;
+		private MemoryHandle _inputBufferHandle;
+		private bool _isDisposed;
+
+		// Note, DeflateStream or the deflater do not try to be thread safe.
+		// The lock is just used to make writing to unmanaged structures atomic to make sure
+		// that they do not get inconsistent fields that may lead to an unmanaged memory violation.
+		// To prevent *managed* buffer corruption or other weird behaviour users need to synchronise
+		// on the stream explicitly.
+		private object SyncLock => this;
+
+		internal Deflater (CompressionAlgorithm algorithm)
+		{
+			DeflateInit (algorithm);
+		}
+
+		~Deflater()
+		{
+			Dispose (false);
+		}
+
+		public void Dispose()
+		{
+			Dispose (true);
+			GC.SuppressFinalize (this);
+		}
+
+		private void Dispose (bool disposing)
+		{
+			if (!_isDisposed) {
+				if (disposing)
+					CompressionStreamStruct.compression_stream_destroy (ref _compression_struct);
```

Comment:
same comment as before, what destroy/release the resources if the GC/finalizer runs this ? `disposed` will be `false` ?!?